driver/power: Add backend driver for EG PMS2 LAN/WLAN

initBasti opened the pull request:
This driver adds support for the Energenie PMS2 LAN & WLAN programmable
power sockets. They are accessed over a web interface and require a
login before accessing the status of the socket or switching them on or off.
A password is required for the login in POST request and this driver
deliberately uses the default password, as the current state of labgrid
doesn't support password encryption, meaning the password would be
stored in plain text.

Signed-off-by: Sebastian Fricke <sebastian.fricke@posteo.net>

Small note about passwords:
From my discussion in this [issue](https://github.com/labgrid-project/labgrid/issues/794) I have gotten the notion, that security in labgrid is performed on the network level, therefore I have decided to keep this module as simple as possible by demanding the default password. If we want to allow the user to change the password, then I believe this patch will become a lot more complex as the NetworkPowerDriver does not support such an attribute and therefore there might be the need for an alternative approach.
I would like to hear your thoughts about this.

What do you use the feature for?
To work with my Energenie programmable power sockets within my personal lab.

How does labgrid benefit as a testing library from the feature?
A wider range of supported devices increases the usefulness of the library for a bigger audience. 

How did you verify the feature works?
I have tested it in my personal lab with the EG_PMS2_LAN device.

If hardware is needed for the feature, which hardware is supported, and which hardware did you test with?
Currently, I know that the device should work for both the EG_PMS2_LAN and EG_PMS2_WLAN devices as both are from the same series, but I only tested with EG_PMS2_LAN. The driver might also work for the EG_PM2 series but I have not tested it (the only difference between the PMS2 and PM2 seems to be surge protection).

**Checklist**
- [x] Documentation for the feature
- [x] Tests for the feature 
- [x] The arguments and description in doc/configuration.rst have been updated
- [ ] Add a section on how to use the feature to doc/usage.rst (should not be required for such a simple module)
- [ ] Add a section on how to use the feature to doc/development.rst (should not be required for such a simple module)
- [x] CHANGES.rst has been updated
- [x] PR has been tested

Emantor reviewed (approved):
LGTM, you surely want to remove the `PORT` variable before merging.

initBasti commented:
@Emantor 
Sorry for the delay, had a busy week. I've removed the line in question, I hope that it is all fine now.
Thank you for the review.

jluebbe commented on labgrid/driver/power/eg_pms2_network.py:
```diff
@@ -0,0 +1,73 @@
+SOCKSTATES_REGEX = str("var\s+sockstates\s+=\s+"
+                       "\[([01]),([01]),([01]),([01])\];")
+
+def login(func):
+    """
+    Decorator automating login and logout from the Energenie web interface
+    """
+    def wrapper(host, port, index, value):
+        base_url = f"http://{host}:{port}"
+
+        login_url = f"{base_url}/login.html"
+        # Login - use the default password 1, because labgrid doesn't support
+        # password encryption, modifying the password doesn't secure the device
+        # as the password would be stored as plain-text.
+        try:
+            response = requests.post(login_url, data={'pw': 1})
+        except requests.exceptions.ConnectionError:
+            raise ExecutionError(f"Device not found at {host} or the network "
+                                 "interface of the device is not active "
+                                 "(press the reset button on the device)")
+        if response.status_code != 200 or not re.search(LOGIN_SUCCESS_REGEX,
+                                                        response.text):
+            raise ExecutionError("Login to Energenie web interface failed")
+
+        result = func(host, port, index, value)
+
+        # Logout
+        response = requests.get(login_url)
+        if response.status_code != 200:
+            raise ExecutionError("Logout from Energenie web interface failed")
+        return result
+    return wrapper
```
Wouldn't two helper functions (login/logout) instead of the decorator be simpler?

initBasti replied:
Hey Jan,

I guess they would be sufficient, I just applied my personal taste in this code and I am not mad if we go another route.
I just felt that...
```python
@login
def power_get(host, port, index):
    ...
    response = requests.get(status_url)
    ...
```
... concentrates more on the action that we are trying to do which is getting the status from the device. The login and logout felt more like a required extra, which can be abstracted out of the specific code into a decorator.
The alternative would be:
```python
def power_get(host, port, index):
    ...
    login(base_url=base_url)
    response = requests.get(status_url)
    logout(base_url=base_url)
    ...
```
Which from my point of view is a lot more verbose, so I prefer the decorator, but again no strong opinions, if simple helper functions fit more into the coding style of labgrid then I will change my code.

Greetings,
Sebastian

initBasti replied:
@jluebbe 
Hey just wanted to check in, I would rebase this commit to the current tree, do you want me to refactor the code as well?

Greetings,
Sebastian

initBasti commented:
Hey @jluebbe @Emantor ,
I have refactored the decorator into two separate functions and documented why the logout functionality is useful. Additionally, I noticed that the change was declared for 3.0 which is already released so I changed it accordingly within the CHANGES.rst file.
I have tested this on my setup and it works without problems.
Greetings,
Sebastian

initBasti commented:
Hey @jluebbe  & @Emantor ,

just wanted to quickly check, is there something left to do for me?

Greetings,
Sebastian

Emantor commented:
> Hey @jluebbe & @Emantor ,
> 
> just wanted to quickly check, is there something left to do for me?

AFAICS no, thanks for the ping.

jluebbe commented on labgrid/driver/power/eg_pms2_network.py:
```diff
@@ -0,0 +1,74 @@
+# Driver has been tested with:
+# * EG_PMS2_LAN
+
+PORT = 80
```
This is needed for the SSH tunnel/proxy. I've made a commit to add it again.

jluebbe reviewed (approved):
Without the `PORT` variable, access via an SSH proxy will not work (see https://github.com/labgrid-project/labgrid/blob/master/labgrid/driver/powerdriver.py#L170). I've added a commit to fix that.

Bastian-Krause commented on labgrid/driver/power/eg_pms2_network.py:
```diff
@@ -0,0 +1,77 @@
+PORT = 80
+
+# The login was successful when we can locate a Log out button
+LOGIN_SUCCESS_REGEX = str("<div class=\"boxmenuitem\"><a href="
+                          "\"login\\.html\">Log Out</a></div>")
```
```suggestion
LOGIN_SUCCESS_REGEX = r'<div class="boxmenuitem"><a href="login\.html">Log Out</a></div>'
```

Bastian-Krause commented on labgrid/driver/power/eg_pms2_network.py:
```diff
@@ -0,0 +1,77 @@
+                          "\"login\\.html\">Log Out</a></div>")
+# Search for a string similar to: "var sockstates = [1,1,1,1];" and create
+# a match group out of the numbers within the square brackets.
+SOCKSTATES_REGEX = str("var\s+sockstates\s+=\s+"
+                       "\[([01]),([01]),([01]),([01])\];")
```
```suggestion
SOCKSTATES_REGEX = r"var\s+sockstates\s+=\s+\[([01]),([01]),([01]),([01])\];"
```

Bastian-Krause reviewed (dismissed):
Looks good to me. I have not tested this, since I don't have the hardware needed.

initBasti commented on labgrid/driver/power/eg_pms2_network.py:
```diff
@@ -0,0 +1,75 @@
+    login(base_url=base_url)
+    response = requests.post(base_url, data={f'cte{index}': value})
+    logout(base_url=base_url)
+    if not response.status_code == 200:
+        raise ExecutionError(f"Switching socket at index {index} "
+                             f"{'on' if value else 'off'} failed")
```
I think in this case it might actually be more descriptive to add the custom message to highlight the state at which the error occurred (index and desired state of the socket).

Bastian-Krause commented on labgrid/driver/power/eg_pms2_network.py:
```diff
@@ -0,0 +1,75 @@
+
+    # Fetch status
+    login(base_url=base_url)
+    response = requests.get(f"{base_url}/energenie.html")
```
This is probably still a good idea, right?

Emantor reviewed (approved):
I rebased and added additional `raise_for_status()` calls.

initBasti commented:
@Emantor Thanks a lot! I was quite busy working myself into my new job, so I delayed finishing this patch... sorry for that!
